### PR TITLE
System.Random performance tuning.

### PR DESCRIPTION
**Pull request by colgreen**
Performance improvements, achieved by operating on a copy of the PRNG state copied into stack variables, and then back onto the heap when done. This benefits all of the various Next() overloads in the region of 10% better performance, and more than doubles performance of NextBytes().

The NextBytes() change means duplicating the PRNG inner logic inside that method, instead of calling out to NextUInt64(). This change alone seems to offer a 2x performance improvement, so is worth considering despite the code duplication. I think the fact that this works suggests that NextUInt64() isn't being inlined(?) (but perhaps I caused that by adding the extra method/stack variables).

I thought this was worth considering given the fairly chunky performance gains. Here are my own BenchmarkDotNet figures (the 10M suffix means a timing for 10 million ops, so divide by 10^7 to get per op time - or just look at the relative times for before and after :)

(for 64bit runtime)

Before
|                 Method |      Mean |    Error |   StdDev |
|----------------------- |----------:|---------:|---------:|
|                Next10M |  25.61 ms | 0.056 ms | 0.050 ms |
|      NextUpperBound10M |  37.55 ms | 0.080 ms | 0.075 ms |
| NextLowerUpperBound10M | 113.47 ms | 0.149 ms | 0.132 ms |
|          NextDouble10M |  31.40 ms | 0.626 ms | 0.792 ms |
|          NextBytes100M |  32.56 ms | 0.078 ms | 0.069 ms |

After
|                 Method |      Mean |    Error |   StdDev |  
|----------------------- |----------:|---------:|---------:|
|                Next10M |  23.78 ms | 0.280 ms | 0.262 ms |
|      NextUpperBound10M |  30.91 ms | 0.325 ms | 0.304 ms |
| NextLowerUpperBound10M | 102.07 ms | 1.436 ms | 1.344 ms |
|          NextDouble10M |  29.51 ms | 0.584 ms | 0.717 ms |
|          NextBytes100M |  14.36 ms | 0.190 ms | 0.177 ms |

/cc @stephentoub

### CODE REVIEW

**Comment by stephentoub** [comment]
Can you share the cited perf tests?

**Comment by colgreen** [comment]
> Can you share the cited perf tests?

https://github.com/colgreen/Redzen/blob/master/Redzen.Benchmarks/Xoshiro256StarStarRandomBenchmark.cs

**Comment by stephentoub** [comment]
Thanks.

>  I think the fact that this works suggests that NextUInt64() isn't being inlined(?)

No, the NextUInt32/64 methods are getting inlined.  It's just that the tight loop ends up having a lot more moves when the fields are involved.
image

**Comment by stephentoub** [comment]
@AndyAyersMS, thoughts on this from a JIT perspective?  Would you expect these kinds of code changes to be necessary, or would you expect the JIT to be able to "get this"?

**Comment by stephentoub** [comment]
> so I think it is technically a change in functionality, and thus something that the JIT cannot do

They're not volatile, so as long as the difference can't be observed by this thread (e.g. if some method were being called that might look at the field value), technically it should be fine.

**Comment by AndyAyersMS** [comment]
> would you expect the JIT to be able to "get this"?

No, this isn't something the JIT does currently. Class field loads and subsequent operations can be CSEd (and hence held in registers), but nothing can promote entire webs involving stores...

**Review by stephentoub — APPROVED** [review]
Thanks.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs [hunk 1]
```diff
@@ -49,16 +49,26 @@ public unsafe XoshiroImpl()
             [MethodImpl(MethodImplOptions.AggressiveInlining)] // small-ish hot path used by a handful of "next" methods
             internal uint NextUInt32()
             {
-                uint result = BitOperations.RotateLeft(_s1 * 5, 7) * 9;
-                uint t = _s1 << 9;
+                uint s0 = _s0;
+                uint s1 = _s1;
+                uint s2 = _s2;
+                uint s3 = _s3;
```
[the remark this answers is not in the document]
Comment (colgreen) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs]: Done. All four nit instance done in single commit.

**Comment by danmosemsft** [comment]
We're missing perf tests for the new API and for the unseeded Random. I'll add something.

**Pull request merged**